core/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java

```diff
@@ -218,84 +203,188 @@ private void startRound() {
-      return now;
-    } else /* now < previousProposerTimestamp */ {
-      /* Our local system time is lagging or the quorum has agreed on a rushing timestamp in the previous round.
-       * We can't use our local time because the proposal would be rejected, hence re-using the previous timestamp
-       * to get as close to the current time (from this node's perspective) as possible. */
-      final var diff = previousProposerTimestamp - now;
-      if (diff > PREVIOUS_ROUND_RUSHING_TIMESTAMP_LOG_THRESHOLD_MS) {
-        log.warn(
-            "Previous round proposer timestamp was greater than the current local system time. "
-                + "This may (but doesn't have to) indicate system clock malfunction. "
-                + "Consider further investigation if this log message appears on a regular basis.");
-      }
-      this.metrics.bft().pacemaker().proposalsWithSubstituteTimestamp().inc();
-      return previousProposerTimestamp;
+  @Override
+  public void preProcessUnsyncedProposalForCurrentOrFutureRound(Proposal proposal) {
+    if (proposal.highQC().getHighestRound().gt(this.highestKnownQcRound)) {
+      this.highestKnownQcRound = proposal.highQC().getHighestRound();
+    }
+    if (proposal.getRound().gt(this.highestReceivedProposalRound)) {
+      this.highestReceivedProposalRound = proposal.getRound();
     }
   }
 
-  private List<RawNotarizedTransaction> getTransactionsForProposal(
-      Round round, QuorumCertificate highestQC) {
-    // If we're at the end of an epoch, we need to generate an empty proposal
-    // - these transactions will get ignored at vertex preparation time anyway
-    // TODO: Remove isEndOfEpoch knowledge from consensus
-    if (highestQC.getProposedHeader().getLedgerHeader().isEndOfEpoch()) {
-      return List.of();
+  @Override
+  public void processProposal(Proposal proposal) {
+    final var proposedVertex = proposal.getVertex().withId(hasher);
+    this.vertexStore.insertVertex(proposedVertex);
+    metrics.bft().successfullyProcessedProposals().inc();
+  }
+
+  @Override
+  public void processBFTUpdate(BFTInsertUpdate update) {
+    log.trace("BFTUpdate: Processing {}", update);
+
+    final var round = update.getHeader().getRound();
+    if (round.lt(this.latestRoundUpdate.getCurrentRound())) {
+      log.trace(
+          "InsertUpdate: Ignoring insert {} for round {}, current round at {}",
+          update,
+          round,
+          this.latestRoundUpdate.getCurrentRound());
+      return;
     }
 
-    final var alreadyExecutedVertices =
-        vertexStore.getPathFromRoot(highestQC.getProposedHeader().getVertexId());
-    final var nextTransactions =
-        proposalGenerator.getTransactionsForProposal(round, alreadyExecutedVertices);
-    this.metrics.bft().pacemaker().proposedTransactions().inc(nextTransactions.size());
-    return nextTransactions;
+    this.maybeLatestInsertUpdate = Optional.of(update);
```

Comment:
I find this `maybeLatestInsertUpdate` confusing - why do we need to store it? Is it only for the future round?

In which case, maybe we separate it into "this.latestInsertUpdateForFutureRound" - only in the case it's for a future round?

And have `this.tryToVoteForTheProposalForTheFirstTime(sendToAll)` become: `this.attemptVoteOnExecutedVertex(BFTInsertUpdate)`, where sendToAll can be read from `this.scheduledRoundTimeoutHasOccurred`?

Similarly, `this.createAndSendTimeoutVoteToAll` and `this.createAndSendTimeoutVote` can be merged into `this.attemptVoteOnExecutedVertex(BFTInsertUpdate)`...

Inside `this.attemptVoteOnExecutedVertex` it also works out whether it should be a timeout vote using `this.scheduledRoundTimeoutHasOccurred` :).

So basically:
* We can vote on a "timeout"/"rejected" vertex without it being a timeout vote (this is why I suggest a rename to "rejection vertex" or "default vertex")
* We add the "timeout" flag to our vote once `this.scheduledRoundTimeoutHasOccurred` and instead send to everyone

Reply:
@dhedey I've refactored it following your ideas (mostly :smile: ), and added some more comments